File: src/utils/documentParser.js
/**
 * Document Parser Service
 * Extracts raw text from PDF, DOCX, TXT, and Markdown files.
 */
const pdfParse = require('pdf-parse');
const mammoth = require('mammoth');
const { marked } = require('marked');
const path = require('path');

// Supported file extensions
const SUPPORTED_EXTENSIONS = ['.pdf', '.docx', '.txt', '.md', '.markdown'];

/**
 * Main entry point - detects file type and delegates to the appropriate parser.
 * @param {Object} file - Multer file object { originalname, buffer, mimetype }
 * @returns {Promise<string>} Extracted plain text
 */
async function parseDocument(file) {
  if (!file || !file.buffer) {
    throw new Error('Invalid file: no buffer provided');
  }

  const ext = path.extname(file.originalname).toLowerCase();

  if (!SUPPORTED_EXTENSIONS.includes(ext)) {
    throw new Error(
      `Unsupported file type: "${ext}". Supported formats: ${SUPPORTED_EXTENSIONS.join(', ')}`
    );
  }

  let text;

  switch (ext) {
    case '.pdf':
      text = await parsePDF(file.buffer);
      break;
    case '.docx':
      text = await parseDOCX(file.buffer);
      break;
    case '.txt':
      text = parseTXT(file.buffer);
      break;
    case '.md':
    case '.markdown':
      text = parseMarkdown(file.buffer);
      break;
    default:
      throw new Error(`No parser available for: ${ext}`);
  }

  // Clean up the extracted text
  text = cleanText(text);

  if (!text || text.trim().length === 0) {
    throw new Error('Document appears to be empty or could not be parsed');
  }

  return text;
}

/**
 * Extract text from a PDF buffer.
 */
async function parsePDF(buffer) {
  try {
    const data = await pdfParse(buffer);
    return data.text;
  } catch (err) {
    throw new Error(`PDF parsing failed: ${err.message}`);
  }
}

/**
 * Extract text from a DOCX buffer.
 */
async function parseDOCX(buffer) {
  try {
    const result = await mammoth.extractRawText({ buffer });
    if (result.messages && result.messages.length > 0) {
      console.warn('DOCX parser warnings:', result.messages);
    }
    return result.value;
  } catch (err) {
    throw new Error(`DOCX parsing failed: ${err.message}`);
  }
}

/**
 * Extract text from a plain text buffer.
 */
function parseTXT(buffer) {
  return buffer.toString('utf-8');
}

/**
 * Extract text from a Markdown buffer (strips markdown syntax).
 */
function parseMarkdown(buffer) {
  const raw = buffer.toString('utf-8');
  // Convert markdown to HTML, then strip HTML tags to get plain text
  const html = marked(raw);
  return html
    .replace(/<[^>]*>/g, ' ')   // strip HTML tags
    .replace(/&[a-z]+;/gi, ' ') // strip HTML entities
    .replace(/\s+/g, ' ')       // collapse whitespace
    .trim();
}

/**
 * Clean extracted text: normalize whitespace, remove control characters.
 */
function cleanText(text) {
  return text
    .replace(/\r\n/g, '\n')         // normalize line endings
    .replace(/[\x00-\x08\x0B\x0C\x0E-\x1F]/g, '') // remove control chars (keep \n, \t)
    .replace(/\n{3,}/g, '\n\n')     // collapse excessive newlines
    .replace(/[ \t]+/g, ' ')        // collapse spaces/tabs
    .trim();
}

module.exports = {
  parseDocument,
  parsePDF,
  parseDOCX,
  parseTXT,
  parseMarkdown,
  cleanText,
  SUPPORTED_EXTENSIONS
};

File: app/utils/document_parser.py
"""
Document Parser
Extracts raw text from PDF, DOCX, TXT, and Markdown files.
"""
import os
import re
from typing import Union

import PyPDF2
import docx
import markdown

SUPPORTED_EXTENSIONS = {".pdf", ".docx", ".txt", ".md", ".markdown"}


async def parse_document(filename: str, content: bytes) -> str:
    """
    Detect file type and extract plain text.

    Args:
        filename: Original filename (used for extension detection).
        content: Raw file bytes.

    Returns:
        Extracted plain text.

    Raises:
        ValueError: Unsupported file type or empty document.
    """
    if not content:
        raise ValueError("Invalid file: no content provided")

    size_kb = len(content) / 1024
    print(f'[Parser] Processing "{filename}" ({size_kb:.1f} KB)')

    ext = os.path.splitext(filename)[1].lower()
    if ext not in SUPPORTED_EXTENSIONS:
        raise ValueError(
            f'Unsupported file type: "{ext}". '
            f"Supported: {', '.join(sorted(SUPPORTED_EXTENSIONS))}"
        )

    if ext == ".pdf":
        text = _parse_pdf(content)
    elif ext == ".docx":
        text = _parse_docx(content)
    elif ext == ".txt":
        text = content.decode("utf-8", errors="replace")
    elif ext in (".md", ".markdown"):
        text = _parse_markdown(content)
    else:
        raise ValueError(f"No parser for: {ext}")

    text = _clean_text(text)
    if not text.strip():
        raise ValueError("Document appears to be empty or could not be parsed")
    return text


# ─── Individual Parsers ───────────────────────────────────

def _parse_pdf(content: bytes) -> str:
    """Extract text from a PDF buffer."""
    import io
    try:
        reader = PyPDF2.PdfReader(io.BytesIO(content))
        pages = [page.extract_text() or "" for page in reader.pages]
        return "\n".join(pages)
    except Exception as e:
        raise ValueError(f"PDF parsing failed: {e}")


def _parse_docx(content: bytes) -> str:
    """Extract text from a DOCX buffer."""
    import io
    try:
        doc = docx.Document(io.BytesIO(content))
        return "\n".join(p.text for p in doc.paragraphs)
    except Exception as e:
        raise ValueError(f"DOCX parsing failed: {e}")


def _parse_markdown(content: bytes) -> str:
    """Convert Markdown to plain text (strip tags)."""
    raw = content.decode("utf-8", errors="replace")
    html = markdown.markdown(raw)
    text = re.sub(r"<[^>]+>", " ", html)       # strip HTML tags
    text = re.sub(r"&[a-z]+;", " ", text, flags=re.I)  # strip entities
    text = re.sub(r"\s+", " ", text)
    return text.strip()


def _clean_text(text: str) -> str:
    """Normalize whitespace and remove control characters."""
    text = text.replace("\r\n", "\n")
    text = re.sub(r"[\x00-\x08\x0B\x0C\x0E-\x1F]", "", text)
    text = re.sub(r"\n{3,}", "\n\n", text)
    text = re.sub(r"[ \t]+", " ", text)
    return text.strip()
